prqlc/prqlc/src/debug.rs

```diff
@@ -0,0 +1,365 @@
+            Ok(())
+        }
+    }
+    let mut io_writer = IoWriter { inner: writer };
+
+    render_html::write_debug_log(&mut io_writer, debug_log)
+}
+
+#[derive(Serialize)]
+pub struct DebugLog {
+    started_at: String,
+    version: String,
+    entries: Vec<DebugEntry>,
+
+    #[serde(skip)]
+    current_stage: Stage,
+
+    #[serde(skip)]
+    suppress: bool,
+}
+
+#[derive(Serialize)]
+struct DebugEntry {
+    stage: Stage,
+    kind: DebugEntryKind,
+}
+
+#[derive(Serialize, AsRefStr)]
+pub enum DebugEntryKind {
+    ReprPrql(crate::SourceTree),
+    ReprLr(lr::Tokens),
+    ReprPr(pr::ModuleDef),
+    ReprPl(pl::ModuleDef),
+    ReprDecl(decl::RootModule),
+    ReprRq(rq::RelationalQuery),
+    ReprSql(String),
+    // TODO: maybe route all log::debug and friends here?
+    // Message(String),
+}
+
+#[derive(Clone, Copy, Serialize, AsRefStr)]
+pub enum Stage {
+    Initial,
+    Parsing(StageParsing),
+    Semantic(StageSemantic),
+    Sql,
+}
+
+impl Stage {
+    fn sub_stage(&self) -> Option<&'_ str> {
+        match self {
+            Stage::Initial => None,
+            Stage::Parsing(s) => Some(s.as_ref()),
+            Stage::Semantic(s) => Some(s.as_ref()),
+            Stage::Sql => None,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Serialize, AsRefStr)]
+pub enum StageParsing {
+    Lexer,
+    Parser,
+}
+
+#[derive(Clone, Copy, Serialize, AsRefStr)]
+pub enum StageSemantic {
+    AstExpand,
+    Resolver,
+    Lowering,
+}
+
+mod render_html {
+    use crate::SourceTree;
+
+    use super::*;
+
+    use std::{
+        collections::HashMap,
+        fmt::{Debug, Result, Write},
+    };
+
+    pub fn write_debug_log<W: Write>(w: &mut W, debug_log: &DebugLog) -> Result {
+        writeln!(w, "<!DOCTYPE html>")?;
+
+        writeln!(w, "<head>")?;
+        writeln!(w, "<title>prqlc debug log {}</title>", debug_log.started_at)?;
+        writeln!(w, "<style>{}</style>", CSS_STYLES)?;
+        writeln!(w, "</head>")?;
+        writeln!(w, "<body>")?;
+
+        // header
+        writeln!(w, "<header>")?;
+        writeln!(w, "<h1>prqlc debug log</h1>")?;
+
+        write_key_values(
+            w,
+            &[
+                ("started_at", &debug_log.started_at),
+                ("version", &debug_log.version),
+            ],
+        )?;
+
+        writeln!(w, "</header>")?;
+
+        // entries
+        writeln!(w, "<div class=\"entries\">")?;
+        for (index, entry) in debug_log.entries.iter().enumerate() {
+            let entry_id = format!("entry-{index}");
+
+            let stage = entry.stage.as_ref().to_lowercase();
+            let substage = entry
+                .stage
+                .sub_stage()
+                .map(|s| s.to_lowercase())
+                .unwrap_or_default();
+            writeln!(w, "<div class=\"entry {stage} {substage}\">",)?;
+
+            writeln!(
+                w,
+                "<input id=\"{entry_id}\" class=\"toggle\" type=\"checkbox\" checked>"
```

Comment:
```suggestion
                r#"<input id="{entry_id}" class="toggle" type="checkbox" checked>"#
```

nitty but this makes it easier to read and I guess write...